### PR TITLE
Refactoring code of SSS and Transmission to be able to share it between material

### PR DESCRIPTION
**Pull request by sebastienlagarde**
This PR move code for SSS and Transmission to SubsurfaceScattering.hlsl and LightEvaluation.hlsl file.
It also rename HasFeatureFlag to HasFlag and move it to common.hlsl

The code is now share between Lit and StackLit and an issue for Transmisison have been fix in StackLit

There is guideline to follow to use this code that are describe in SubsurfaceScattering.hlsl

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+// This files include various helper function to easily setup SSS and transmission inside a material. It require that the material follow naming convention
+// User can request either SSS function, or Transmission, or both, they need to define INCLUDE_SUBSURFACESCATTERING and/or INCLUDE_TRANSMISSION
+// And define the lighting model to use for transmission. By default it is Lambert. For Disney: TRANSMISSION_DISNEY_DIFFUSE_BRDF
+// Also user need to be sure that upper 16bit of bsdfData.materialFeatures are not used
+
+// Additional bits set in 'bsdfData.materialFeatures' to save registers and simplify feature tracking.
+#define MATERIAL_FEATURE_SSS_TRANSMISSION_START (1 << 16) // It should be safe to start these flags
+
+#define MATERIAL_FEATURE_FLAGS_SSS_OUTPUT_SPLIT_LIGHTING         ((MATERIAL_FEATURE_SSS_TRANSMISSION_START) << 0)
+#define MATERIAL_FEATURE_FLAGS_SSS_TEXTURING_MODE_OFFSET FastLog2((MATERIAL_FEATURE_SSS_TRANSMISSION_START) << 1) // 2 bits
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Add a comment that the texturing mode is 2 bits, not the offset

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+#ifdef INCLUDE_SUBSURFACESCATTERING
+// Assume that bsdfData.diffusionProfile is init
+void FillMaterialSSS(uint diffusionProfile, float subsurfaceMask, inout BSDFData bsdfData)
+{
+    bsdfData.diffusionProfile = diffusionProfile;
+    bsdfData.fresnel0 = _TransmissionTintsAndFresnel0[diffusionProfile].a;
+    bsdfData.subsurfaceMask = subsurfaceMask;
+    bsdfData.materialFeatures |= MATERIAL_FEATURE_FLAGS_SSS_OUTPUT_SPLIT_LIGHTING;
+    bsdfData.materialFeatures |= GetSubsurfaceScatteringTexturingMode(bsdfData.diffusionProfile) << MATERIAL_FEATURE_FLAGS_SSS_TEXTURING_MODE_OFFSET;
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Should we merge this file with SubsurfaceScattering.hlsl?

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+bool ShouldOutputSplitLighting(BSDFData bsdfData)
+{
+    return HasFeatureFlag(bsdfData.materialFeatures, MATERIAL_FEATURE_FLAGS_SSS_OUTPUT_SPLIT_LIGHTING);
+}
+
+float3 SSSGetModifiedDiffuseColor(BSDFData bsdfData)
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Let's not start naming things with SSS, that's not how we've done is so far.

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+{
+    // Subsurface scattering mdoe
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Typo

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 3 above)
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Do we really need a function for this? We already have ApplySubsurfaceScatteringTexturingMode()?

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    bsdfData.transmittance = ComputeTransmittanceJimenez(_HalfRcpVariancesAndWeights[diffusionProfile][0].rgb,
+                                                         _HalfRcpVariancesAndWeights[diffusionProfile][0].a,
+                                                         _HalfRcpVariancesAndWeights[diffusionProfile][1].rgb,
+                                                         _HalfRcpVariancesAndWeights[diffusionProfile][1].a,
+                                                         _TransmissionTintsAndFresnel0[diffusionProfile].rgb,
+                                                         bsdfData.thickness);
+#endif
+}
+
+#ifdef HAS_LIGHTLOOP
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Why do we need this?

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+//   for the thickness) to compute the transmittance;
+// - we multiply the transmitted radiance by the transmittance.
+float3 EvaluateTransmission(BSDFData bsdfData, float3 transmittance, float NdotL, float NdotV, float LdotV, float attenuation)
+{
+    // Apply wrapped lighting to better handle thin objects at grazing angles.
+    float wrappedNdotL = ComputeWrappedDiffuseLighting(-NdotL, SSS_WRAP_LIGHT);
+
+    // Apply BSDF-specific diffuse transmission to attenuation. See also: [SSS-NOTE-TRSM]
+    // We don't multiply by 'bsdfData.diffuseColor' here. It's done only once in PostEvaluateBSDF().
+#ifdef TRANSMISSION_DISNEY_DIFFUSE_BRDF
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: We should use the BRDF macro to have visuals consistent with the BRDF

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    attenuation *= DisneyDiffuse(NdotV, max(0, -NdotL), LdotV, bsdfData.perceptualRoughness);
+#else
+    attenuation *= Lambert();
+#endif
+
+    float intensity = attenuation * wrappedNdotL;
+    return intensity * transmittance;
+}
+
+void PreEvaluateLightTransmission(float NdotL, BSDFData bsdfData, inout float3 normalWS, inout float contactShadowIndex)
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Why is contactShadowIndex a float?

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    // However we flip the normal for the bias (and the NdotL test) and disable contact shadow
+    if (HasFeatureFlag(bsdfData.materialFeatures, MATERIAL_FEATURE_FLAGS_TRANSMISSION_MODE_THIN_THICKNESS) && NdotL < 0)
+    {
+        //  Disable shadow contact in case of transmission and backface shadow
+        normalWS = -normalWS;
+        contactShadowIndex = -1;
+    }
+}
+
+float3 PostEvaluateLightTransmission(   ShadowContext shadowContext, PositionInputs posInput, float distFrontFaceToLight,
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Post Evaluate cannot include Evaluation itself... Can we merge it into the function above as PreEvaluateTransmission? Then we perform EvaluateTransmission outside.

File: com.unity.render-pipelines.core/CoreRP/ShaderLibrary/CommonMaterial.hlsl [hunk 9]
```diff
@@ -236,4 +236,14 @@ real3 LerpWhiteTo(real3 b, real t)
     return real3(oneMinusT, oneMinusT, oneMinusT) + b * t;
 }
 
+// ----------------------------------------------------------------------------
+// Helper
+// ----------------------------------------------------------------------------
+
+// Simple function to test a bitfield
+bool HasFeatureFlag(uint featureFlags, uint flag)
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.core/CoreRP/ShaderLibrary/CommonMaterial.hlsl]: Should probably call it HasFlag as the function doesn't really have anything to do with features. Especially now that it's in core. Should it be in Common.hlsl?

Comment (sebastienlagarde) [reply com.unity.render-pipelines.core/CoreRP/ShaderLibrary/CommonMaterial.hlsl]: good point

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 2 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Maybe, we can discuss about it

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 3 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: yes, I prefer to abstract what access to flag like MATERIAL_FEATURE_FLAGS_SSS_TEXTURING_MODE_OFFSET

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 5 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: else it don't compile as several file like out debug file include lit.hlsl without having the lightloop part. And function inside HAS_LIGHTLOOP guard are for the lightloop part

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 6 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: What are BRDF macro?

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 7 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: good catch :)

File: com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl (hunk 8 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/HDRP/Material/SubsurfaceScattering/SubsurfaceScatteringUtils.hlsl]: Let's discuss

File: com.unity.render-pipelines.high-definition/HDRP/Lighting/LightEvaluation.hlsl [hunk 10]
```diff
@@ -249,3 +249,131 @@ void EvaluateLight_EnvIntersection(float3 positionWS, float3 normalWS, EnvLightD
...
+// Helper functions to use Transmission with a material
+// ----------------------------------------------------------------------------
+// For EvaluateTransmission.hlsl file it is required to define a BRDF for the transmission. Defining USE_DIFFUSE_LAMBERT_BRDF use Lambert, otherwise it use Disneydiffuse
+
+#ifdef MATERIAL_INCLUDE_TRANSMISSION
+
+// This function return transmittance to provide to EvaluateTransmission
+float3 PreEvaluatePunctualLightTransmission(LightLoopContext lightLoopContext, PositionInputs posInput, float distFrontFaceToLight,
+                                            float NdotL, float3 L, LightData lightData, BSDFData bsdfData,
+                                            inout float3 normalWS, inout int contactShadowIndex)
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/HDRP/Lighting/LightEvaluation.hlsl]: contactShadowIndex is a part of LightData. We already pass LightData, so just pass it as inout?

**Pull request merged**